jobmon_client/src/jobmon/client/task_group.py

```diff
@@ -0,0 +1,227 @@
+
+    def get_task(self, **kwargs) -> Task:
+        """Raises an error if it doesn't uniquely identify a task.
+
+        kwargs can be task arguments or attributes.
+        """
+        tasks = list(self.get_subset(**kwargs))
+
+        if len(tasks) == 0:
+            raise ValueError("No matching tasks in this TaskGroup.")
+
+        if len(tasks) > 1:
+            raise ValueError("Provided labels do not uniquely identify a task.")
+
+        return tasks[0]
+
+    def interleave_upstream(
+        self,
+        upstream_group: TaskGroup,
+        dependency_specification: Dict[str, str],
+        subsetter: Optional[Dict] = None,
+        upstream_subsetter: Optional[Dict] = None,
+    ):
+        """Set interleaving dependencies between TaskGroups via label value matching.
+
+        Dependencies are set between tasks where the values in all the specified labels match
+        and the interleaved tasks fulfill any specified subset criteria.
+
+        Args:
+            upstream_group: The task group to interleave dependencies from.
+            dependency_specification: A dictionary mapping label names between groups. Keys
+                are labels on tasks in this group, and values are labels in tasks in the
+                upstream_group.
+            subsetter: A dictionary used to subset the tasks of the current TaskGroup by
+                filtering label keys and values based on the supplied dict.
+            upstream_subsetter: A dictionary used to subset the tasks of the upstream TaskGroup
+                by filtering label keys and values based on the supplied dict.
+        """
+
+        # get the set of tasks for this TaskGroup based on the provided subsetter
+        if subsetter is not None:
+            tasks = self.get_subset(**subsetter)
+        else:
+            tasks = self.tasks
+
+        # get the set of tasks for this TaskGroup based on the provided subsetter
+        if upstream_subsetter is not None:
+            upstream_tasks = upstream_group.get_subset(**upstream_subsetter)
+        else:
+            upstream_tasks = upstream_group.tasks
+        upstream_label_map = get_label_map(
+            upstream_tasks, list(dependency_specification.values())
+        )
+
+        matched_upstreams = set()
+        matched_downstreams = set()
+        for task in tasks:
+            matches = upstream_tasks.copy()
+            for key, upstream_key in dependency_specification.items():
+                if key in task.labels and upstream_key in upstream_label_map:
+                    label_value = task.labels[key]
+                    matches.intersection_update(
+                        upstream_label_map[upstream_key][label_value]
+                    )
+                else:
+                    matches = set()
+            if matches:
+                matched_upstreams.update(matches)
+                matched_downstreams.add(task)
+                task.add_upstreams(list(matches))
```

Comment:
Since you're calling `set(tasks)` in the init method, will upstreams set in this method be propagated to the input iterable of tasks? 

I think since tasks are mutable objects and no explicit deepcopy is called then mutations to a task in any data structure will persist globally, which is probably the expected behavior, but it might be worth a double check.

```
workflow.add_tasks(upstreams)
workflow.add_tasks(downstreams)
tg = TaskGroup(downstreams)
tg.interleave_upstream(TaskGroup(upstreams))
workflow.run()  # downstream tasks on the workflow should have the upstream tasks attribute properly set
```

Reply:
yeah they should all be propagated to the tasks, we aren't copying the actual tasks, we're using the same task objects. At least some of the tests are structured in a way that demonstrates this I think